lib/system/strs_v3.nim

```diff
@@ -0,0 +1,229 @@
+  ## TODO: fixme: inject conversions somehwere
+  if s.len == 0: result = cstring""
+  else: result = cast[cstring](unsafeAddr s.p.data)
+
+proc appendString(dest: var NimStringV3; src: NimStringV3) {.compilerproc, inline.} =
+  if src.len > 0:
+    copyMem(unsafeAddr dest.p.data[dest.len], unsafeAddr src.p.data[0], src.len)
+    inc dest.len, src.len
+
+proc appendChar(dest: var NimStringV3; c: char) {.compilerproc, inline.} =
+  dest.p.data[dest.len] = c
+
+proc rawNewString(space: int): NimStringV3 {.compilerproc.} =
+  # this is also 'system.newStringOfCap'.
+  if space <= 0:
+    result = NimStringV3(len: 0, p: nil)
+  else:
+    var p = allocPayload(space)
+    p.cap = space
+    result = NimStringV3(len: 0, p: p)
+
+proc mnewString(len: int): NimStringV3 {.compilerproc.} =
+  if len <= 0:
+    result = NimStringV3(len: 0, p: nil)
+  else:
+    var p = allocPayload0(len)
+    p.cap = len
+    result = NimStringV3(len: len, p: p)
+
+proc setLengthStrV2(s: var NimStringV3, newLen: int) {.compilerRtl.} =
+  if newLen == 0:
+    discard "do not free the buffer here, pattern 's.setLen 0' is common for avoiding allocations"
+  else:
+    if isLiteral(s):
+      let oldP = s.p
+      s.p = allocPayload(newLen)
+      s.p.cap = newLen
+      if s.len > 0:
+        copyMem(unsafeAddr s.p.data[0], unsafeAddr oldP.data[0], min(s.len, newLen))
+        if newLen > s.len:
+          zeroMem(cast[pointer](addr s.p.data[s.len]), newLen - s.len)
+        # else:
+        #   s.p.data[newLen] = '\0'
+      else:
+        zeroMem(cast[pointer](addr s.p.data[0]), newLen)
+    elif newLen > s.len:
+      let oldCap = s.p.cap
+      if newLen > oldCap:
+        let newCap = max(newLen, resize(oldCap))
+        s.p = reallocPayload0(s.p -% sizeof(NimStrPayloadBase), oldCap, newCap)
+        s.p.cap = newCap
+
+  s.len = newLen
+
+proc nimAsgnStrV2(a: var NimStringV3, b: NimStringV3) {.compilerRtl.} =
+  if a.p == b.p: return
+  if isLiteral(b):
+    # we can shallow copy literals:
+    frees(a)
+    a.len = b.len
+    a.p = b.p
+  else:
+    if isLiteral(a) or a.p.cap < b.len:
+      # we have to allocate the 'cap' here, consider
+      # 'let y = newStringOfCap(); var x = y'
+      # on the other hand... These get turned into moves now.
+      frees(a)
+      a.p = allocPayload(b.len)
+      a.p.cap = b.len
+    a.len = b.len
+    copyMem(unsafeAddr a.p.data[0], unsafeAddr b.p.data[0], b.len)
+
+proc nimPrepareStrMutationImpl(s: var NimStringV3) =
+  let oldP = s.p
+  # can't mutate a literal, so we need a fresh copy here:
+  s.p = allocPayload(s.len)
+  s.p.cap = s.len
+  copyMem(unsafeAddr s.p.data[0], unsafeAddr oldP.data[0], s.len)
+
+proc nimPrepareStrMutationV2(s: var NimStringV3) {.compilerRtl, inl.} =
+  if isLiteral(s):
+    nimPrepareStrMutationImpl(s)
+
+proc prepareMutation*(s: var string) {.inline.} =
+  # string literals are "copy on write", so you need to call
```

Comment:
why not remove the string literal copy-on-write thing altogether? literals are short and typically unimportant for performance but they are give an outsize implementation space cost here..  there any benchmark showing an actual difference?